File: Lab 2/lab2.erl
-module(lab2).

%
% -- Public --
%

-export([main/2, test/0, test2/0, test3/0, test4/0]).

% Checking for positive integers then creating the lists
main(Length, Spacer) when is_integer(Length), Length >= 0,
                          is_integer(Spacer), Spacer >= 0 -> 
    
    GeneratedLists = generateLists(Length, Spacer), % Creates variable that can be returned
    formatLists(GeneratedLists), % Formats the list to ignore the ascii value pattern matching
    GeneratedLists; % Returns the list of lists from the function

% Character and negative number handling
main(_,_) -> 
    {error, "Invalid input: Please enter a non-negative integer."}.


% The orginal test case from the lab page.
test() -> 
    ExpectedOutput = [
    [14,28,42,56,70,84],
    [13,27,41,55,69,83],
    [12,26,40,54,68,82],
    [11,25,39,53,67,81],
    [10,24,38,52,66,80],
    [9,23,37,51,65,79],
    [8,22,36,50,64,78],
    [7,21,35,49,63,77],
    [6,20,34,48,62,76],
    [5,19,33,47,61,75],
    [4,18,32,46,60,74],
    [3,17,31,45,59,73],
    [2,16,30,44,58,72],
    [1,15,29,43,57,71]],

    Result = main(6,14),

    % Better test case format
    case Result of 
        ExpectedOutput ->
            io:format("test passed~n");
        _ -> 
            io:format("test failed~n")
    end.

% Test case with different integer values.
test2() ->
    ExpectedOutput = [
    [10,20,30,40,50],
    [9,19,29,39,49],
    [8,18,28,38,48],
    [7,17,27,37,47],
    [6,16,26,36,46],
    [5,15,25,35,45],
    [4,14,24,34,44],
    [3,13,23,33,43],
    [2,12,22,32,42],
    [1,11,21,31,41]],

    Result = main(5,10),
    
    % Test case output
    case Result of 
        ExpectedOutput ->
            io:format("test passed~n");
        _ -> 
            io:format("test failed~n")
    end.


% Test case with a negative integer.
test3() ->
    ExpectedOutput = {error, "Invalid input: Please enter a non-negative integer."},

    Result = main(-5,10),
    
    % Test case output
    case Result of 
        ExpectedOutput ->
            io:format("test passed~n");
        _ -> 
            io:format("test failed~n")
    end.

% Test case with both a negative integer and a character.
test4() ->
    ExpectedOutput = {error, "Invalid input: Please enter a non-negative integer."},

    Result = main(-5,"e"),
    
    % Test case output
    case Result of 
        ExpectedOutput ->
            io:format("test passed~n");
        _ -> 
            io:format("test failed~n")
    end.

% -- Private --

% Debugged and slightly altared by chatgpt
% The inner list comprehension generates the N-length sequences spaced by M.
% The outer list comprehension generates the M lists.

generateLists(Length, Spacer) -> [ [X + (Spacer*Y) || Y <- lists:seq(0, Length-1)] || X <- lists:seq(Spacer, 1, -1) ].


% Formats the list to avoid the pattern matching of ascii values in the lists.
% Debugged by chatgpt to have it just format the lists instead of returning the list as well.

formatLists([]) -> ok;
formatLists([HeadList | TailLists]) -> 
    io:format("~w~n", [HeadList]), formatLists(TailLists).

File: Lab 2/lab2.java
import java.util.ArrayList;
import java.util.List;

public class lab2{

    // creates the lists using the 2 values taken in and returns the list of the lists
    public static List<List<Integer>> generateLists(int length, int spacer){

        if(!(length >= 0 && length == (int)length) || !(spacer > 0 && spacer == (int)spacer)){
            throw new IllegalArgumentException("Error: Please input a non-negative integer");
        }
        
        List<List<Integer>> lists = new ArrayList<>();

        // calculates and populates the values in each list
        for(int i = spacer; i >= 1; i--){
            List<Integer> list = new ArrayList<>();
            for(int j = 0; j < length; j++){
                list.add(i + (spacer * j));
            }
            lists.add(list);
        }
        return lists;
    }

    
    public static void main(String[] args){
        
        System.out.println();
        System.out.println("-----------------------------------------");
        System.out.println("Test Case 1: example from lab page");
        System.out.println();

        try {
            
            // Test case for the example on the lab page
            List<List<Integer>> lists = generateLists(6, 14);

            for(List<Integer> list : lists){
                System.out.println(list);
            }
        } catch (IllegalArgumentException exception) {
            
            System.out.println(exception.getMessage());
        }
        
        System.out.println();
        System.out.println("-----------------------------------------");
        System.out.println("Test Case 2: More expected values");
        System.out.println();

        try{
            
            // Another test case with expected values
            List<List<Integer>> lists2 = generateLists(5, 10);

            for(List<Integer> list : lists2){
                System.out.println(list);
            }
        }
        catch(IllegalArgumentException exception){
            System.out.println(exception.getMessage());
        }

        System.out.println();
        System.out.println("-----------------------------------------");
        System.out.println("Test Case 3: negative integer handling");
        System.out.println();
 
        try{
            
            // Test case for negative integers
            List<List<Integer>> lists3 = generateLists(-5,6);

            for(List<Integer> list : lists3){
                System.out.println(list);
            }
        }
        catch(IllegalArgumentException exception){
            System.out.println(exception.getMessage());
        }

        System.out.println();
        System.out.println("-----------------------------------------");
        System.out.println("Test Case 4: Wanted to use a character to test my handling"); 
        System.out.println("but it wouldn't compile bc its not an integer obviously"); 
        System.out.println();

        try{
            
            /* 
            Wanted to test for more unexpected inputs (ie. char / String) but obviously it wouldn't 
            compile because the generateLists method takes in only ints as its parameters
            */ 
            List<List<Integer>> lists4 = generateLists(-7, 8);

            for(List<Integer> list : lists4){
                System.out.println(list);
            }
        }
        catch(IllegalArgumentException exception){
            System.out.println(exception.getMessage());
        }

        System.out.println();
        System.out.println("-----------------------------------------");
        System.out.println();
    }
}